Enabling Support for Conditional Multi-Part Upload

x-INFiN1TY-x opened the pull request:
<html><head></head><body><h1>Enabling Support for Conditional Multi-Part Upload (#18093)</h1>
<p><strong>At a Glance:</strong><br>
Implement both synchronous and asynchronous conditional-upload methods in <code inline="">S3BlobContainer</code>, backed by logical functions in <code inline="">AsyncTransferManager</code></p>

Method | Location | Purpose/Functionality
-- | -- | --
executeMultipartUploadConditionally | S3BlobContainer | Conditional multipart upload (sync)
asyncBlobUploadConditionally | S3BlobContainer | Conditional streamed upload (async)
uploadObjectConditionally | AsyncTransferManager | Core async conditional upload implementation
Helpers in AsyncTransferManager |   | uploadInOneChunkConditionally
Helpers in AsyncTransferManager |   | uploadInPartsConditionally
Helpers in AsyncTransferManager |   | doUploadInPartsConditionally
Helpers in AsyncTransferManager |   | handleConditionalException


<hr>
<h2>Summary of Changes</h2>
<ul>
<li>
<p><strong>Add</strong> <code inline="">executeMultipartUploadConditionally(...)</code> to <code inline="">S3BlobContainer</code></p>
</li>
<li>
<p><strong>Add</strong> <code inline="">asyncBlobUploadConditionally(...)</code> to <code inline="">S3BlobContainer</code></p>
</li>
<li>
<p><strong>Implement</strong> <code inline="">uploadObjectConditionally(...)</code> in <code inline="">AsyncTransferManager</code></p>
</li>
<li>
<p><strong>Introduce</strong> helper routines: <code inline="">uploadInOneChunkConditionally</code>, <code inline="">uploadInPartsConditionally</code>, <code inline="">doUploadInPartsConditionally</code>, and <code inline="">handleConditionalException</code></p>
</li>
<li>
<p><strong>Define</strong> <code inline="">ConditionalWriteOptions</code> to express “If-Match” and “If-None-Match” conditions</p>
</li>
<li>
<p><strong>Define</strong> <code inline="">ConditionalWriteResponse</code> to wrap returned ETag or propagate failures</p>
</li>
<li>
<p><strong>Extend</strong> tests to cover conditional success, precondition failures, data integrity, and size boundaries</p>
</li>
</ul>
<hr>
<h2>Core Workflow</h2>
<ol>
<li>
<p><strong>Validate</strong> part size (≥ 5 MB) and total upload size (≤ 5 TB).</p>
</li>
<li>
<p><strong>Initialize</strong> upload by creating a multipart request with bucket, key, ACL, storage class, metadata, and SSE (AES256 or KMS).</p>
</li>
<li>
<p><strong>Transfer Data</strong></p>
<ul>
<li>
<p><strong>Sync</strong>: iterate parts sequentially using AWS SDK, buffering if retries are enabled.</p>
</li>
<li>
<p><strong>Async</strong>: delegate to <code inline="">AsyncTransferManager.uploadObjectConditionally</code>, which chooses single-chunk or multipart flow via its helper methods.</p>
</li>
</ul>
</li>
<li>
<p><strong>Complete Conditionally</strong> by setting either <code inline="">If-Match</code> or <code inline="">If-None-Match: *</code> on the complete-upload request based on <code inline="">ConditionalWriteOptions</code>.</p>
</li>
<li>
<p><strong>Respond</strong> with <code inline="">ConditionalWriteResponse.success(etag)</code> on success or translate failures:</p>
<ul>
<li>
<p><strong>HTTP 412</strong> → <code inline="">OpenSearchException("stale_primary_shard")</code> + abort multipart.</p>
</li>
<li>
<p><strong>Other errors</strong> → <code inline="">IOException</code> (sync) or failed <code inline="">CompletableFuture</code>/listener (async) + abort if uploadId exists.</p>
</li>
<li>
<p><strong>Abort failures</strong> are logged but do not mask the original exception.</p>
</li>
</ul>
</li>
</ol>
<hr>
<h2>Test Coverage</h2>
<p><strong>executeMultipartUploadConditionally Test Suite</strong></p>
<ul>
<li>
<p><code inline="">testExecuteMultipartUploadConditionallyWithEtagMatchSuccess</code></p>
</li>
<li>
<p><code inline="">testExecuteMultipartUploadConditionallyWithMetadataAndSSE</code></p>
</li>
<li>
<p><code inline="">testExecuteMultipartUploadConditionallyContentIntegrity</code></p>
</li>
<li>
<p><code inline="">testExecuteMultipartUploadConditionallySizeValidation</code></p>
</li>
<li>
<p><code inline="">testExecuteMultipartUploadConditionallyPreconditionFailed</code></p>
</li>
</ul>
<p><strong>AsyncTransferManager Conditional Upload Tests</strong></p>
<ul>
<li>
<p><code inline="">testConditionalOneChunkUpload</code></p>
</li>
<li>
<p><code inline="">testConditionalOneChunkUploadPreconditionFailed</code></p>
</li>
<li>
<p><code inline="">testConditionalOneChunkUploadCorruption</code></p>
</li>
<li>
<p><code inline="">testConditionalMultipartUploadPreconditionFailed</code></p>
</li>
<li>
<p><code inline="">testConditionalMultipartUploadCorruption</code></p>
</li>
</ul>
<p><strong>Container-Level and Retry Tests</strong></p>
<ul>
<li>
<p><code inline="">testFailureWhenLargeFileRedirectedConditional</code></p>
</li>
<li>
<p><code inline="">testLargeFileRedirectedConditional</code></p>
</li>
<li>
<p><code inline="">testLargeFileRedirectedConditionalPreconditionFailed</code></p>
</li>
<li>
<p><code inline="">testLargeFileRedirectedConditionalConflict</code></p>
</li>
<li>
<p><code inline="">testWriteBlobByStreamsConditionalFailure</code></p>
</li>
</ul>

---

### Related Issue
Concerned RFC : [RFC #17763](https://github.com/opensearch-project/OpenSearch/issues/17763)
Parent Meta Issue : [Meta #17859](https://github.com/opensearch-project/OpenSearch/issues/17859)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### Flow Diagram for executeMultipartUploadConditionally:

![Mermaid Chart - Create complex, visual diagrams with text  A smarter way of creating diagrams -2025-05-29-085837](https://github.com/user-attachments/assets/94efdf1b-493e-4c31-8a35-2dc34feda5f9)

tqranjan committed:
git commit -m "UTs for : executeMultipartUploadIfEtagMatches"

tqranjan committed:
git commit -m "Implementation for : executeMultipartUploadIfEtagMatches"

x-INFiN1TY-x commented:
Flaky Test :: https://github.com/opensearch-project/OpenSearch/issues/17540 | https://github.com/opensearch-project/OpenSearch/issues/17551
- classMethod – org.opensearch.repositories.s3.S3BlobContainerRetriesTests

tqranjan committed:
Adds a constant HTTP_STATUS_PRECONDITION_FAILED with value 412

Signed-off-by: Tanishq Ranjan <tqranjan@amazon.com>

Bukhtawar reviewed (commented):
We also support async multipart upload(`AsyncTransferManager` has the logic). Do you want to extend that capability as well? 
cc: @ashking94

Bukhtawar commented on plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -509,6 +514,169 @@ private String buildKey(String blobName) {
         return keyPath + blobName;
     }
 
+    public void executeMultipartUploadIfEtagMatches(
```
Where is this code getting consumed?

x-INFiN1TY-x replied:
![Diag](https://github.com/user-attachments/assets/2992d8dd-7696-4da4-ac41-16d29d12cc1d)

Sorry missed this,  The `executeMultipartUploadIfEtagMatches` method (now renamed to `executeMultipartUploadConditionally`) is meant to be consumed via the `writeBlobWithMetadataConditionally` and `writeBlobConditionally` methods in the generic blob store interface. These interfaces are introduced in [this PR currently under review](https://github.com/opensearch-project/OpenSearch/pull/18064).

ashking94 commented:
> We also support async multipart upload(`AsyncTransferManager` has the logic). Do you want to extend that capability as well? cc: @ashking94

Check https://github.com/opensearch-project/OpenSearch/pull/18064#pullrequestreview-2814267845. I had the same question. Will let this be addressed at both places.

tqranjan committed:
UploadRequest : Add ConditionalWriteOptions

tqranjan committed:
Add ConditionalWrite utility for conditional blob store writes

Introduces ConditionalWriteOptions and ConditionalWriteResponse classes to support conditional write operations in BlobContainer implementations.

tqranjan committed:
Adds logic to Upload an object to S3 conditionally using the async client

tqranjan committed:
UTs for async Conditional upload logic

tqranjan committed:
Test Suite for ASync Conditional upload process

tqranjan committed:
S3 Blobcontainer Implementation for Async Conditional upload

tqranjan committed:
Javadoc for builder

tqranjan committed:
asyncBlobUploadConditionally : Interfaces for AsyncMultiStreamBlobContainer & AsyncMultiStreamEncryptedBlobContainer

tqranjan committed:
Log fix

x-INFiN1TY-x commented:
Flaky Test :: https://github.com/opensearch-project/OpenSearch/issues/17551

classMethod – org.opensearch.repositories.s3.S3BlobContainerRetriesTests

x-INFiN1TY-x committed:
Merge branch 'main' into pr-branch-conditionalMultipartUpload

Signed-off-by: Tanishq Ranjan <111151372+x-INFiN1TY-x@users.noreply.github.com>

tqranjan committed:
SpotlessApply post Conflict resolution

tqranjan committed:
Conflict resolution in executeMultipartUploadConditionally

tqranjan committed:
Conflict resolution in executeMultipartUploadConditionally's UTs

tqranjan committed:
Conflict resolution in S3BlobContainerMockClientTests & AsyncTransferManagerTests

x-INFiN1TY-x commented:
Flaky Test :: https://github.com/opensearch-project/OpenSearch/issues/14293

ResourceAwareTasksTests #14293

astute-decipher commented on server/src/main/java/org/opensearch/common/blobstore/ConditionalWrite.java:
```diff
@@ -0,0 +1,171 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore;
```
Can we remove this from this PR once the parent PR is merged : #18064

astute-decipher commented on plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -209,6 +216,124 @@ public void writeBlobWithMetadata(
 
+    @Override
+    public void asyncBlobUploadConditionally(
+        WriteContext writeContext,
```
Lets refactor and avoid code duplication

tqranjan committed:
S3BlobContainer Refactor

tqranjan committed:
AsyncTransferManager Refactor

tqranjan committed:
removed stale_primary_shard : generic rewording

tqranjan committed:
AsyncTransferManagerTests.java refactor

shourya035 commented on plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -236,19 +262,33 @@ public void asyncBlobUpload(WriteContext writeContext, ActionListener<Void> comp
-                    );
-                    completionListener.onResponse(null);
+                    if (options != null) {
+                        executeMultipartUploadConditionally(
```
Let's see if this can be refactored a bit differently. A common `prepareS3MultiPartUpload` can be used to generate the request payload and `options` can be applied to that conditionally here. The flow would look something like:

```
if (options != null) {
  request = prepareS3MultiPartUpload();
  applyOptionsToRequest(request)
} else {
  prepareS3MultiPartUpload();
}
```

Improves code re-usability

shourya035 commented on plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -521,6 +612,177 @@ private String buildKey(String blobName) {
         return keyPath + blobName;
     }
 
+    public void executeMultipartUploadConditionally(
```
Continuing from https://github.com/opensearch-project/OpenSearch/pull/18093/files#r2702258305

We can refactor the common code from `executeMultipartUploadConditionally` and `executeSingleUpload` to a single function

shourya035 commented on plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncTransferManager.java:
```diff
@@ -447,6 +483,155 @@ private void uploadInOneChunk(
+
+        if (cause instanceof S3Exception s3e) {
+            if (s3e.statusCode() == HTTP_STATUS_PRECONDITION_FAILED) {
+                returnFuture.completeExceptionally(
```
We should have a `ConditionalWriteResponse.failed` as well? Instead of throwing a S3Exception from here directly, we should wrap that around the `ConditionalWriteResponse` class?
